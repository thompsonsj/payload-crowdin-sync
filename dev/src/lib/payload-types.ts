/* tslint:disable */
/* eslint-disable */
/**
 * This file was automatically generated by Payload.
 * DO NOT MODIFY IT BY HAND. Instead, modify your source Payload config,
 * and re-run `payload generate:types` to regenerate this file.
 */

export interface Config {
  collections: {
    categories: Category;
    'multi-rich-text': MultiRichText;
    'localized-posts': LocalizedPost;
    media: Media;
    'nested-field-collection': NestedFieldCollection;
    policies: Policy;
    posts: Post;
    'localized-posts-with-condition': LocalizedPostsWithCondition;
    tags: Tag;
    users: User;
    'crowdin-files': CrowdinFile;
    'crowdin-collection-directories': CrowdinCollectionDirectory;
    'crowdin-article-directories': CrowdinArticleDirectory;
    'payload-preferences': PayloadPreference;
    'payload-migrations': PayloadMigration;
  };
  globals: {
    'localized-nav': LocalizedNav;
    nav: Nav;
    statistics: Statistic;
  };
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "categories".
 */
export interface Category {
  id: string;
  name?: string | null;
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "multi-rich-text".
 */
export interface MultiRichText {
  id: string;
  field_0?:
    | {
        [k: string]: unknown;
      }[]
    | null;
  field_1?:
    | {
        [k: string]: unknown;
      }[]
    | null;
  field_2?:
    | {
        [k: string]: unknown;
      }[]
    | null;
  field_3?:
    | {
        [k: string]: unknown;
      }[]
    | null;
  field_4?:
    | {
        [k: string]: unknown;
      }[]
    | null;
  field_5?:
    | {
        [k: string]: unknown;
      }[]
    | null;
  field_6?:
    | {
        [k: string]: unknown;
      }[]
    | null;
  field_7?:
    | {
        [k: string]: unknown;
      }[]
    | null;
  field_8?:
    | {
        [k: string]: unknown;
      }[]
    | null;
  field_9?:
    | {
        [k: string]: unknown;
      }[]
    | null;
  field_10?:
    | {
        [k: string]: unknown;
      }[]
    | null;
  syncTranslations?: boolean | null;
  syncAllTranslations?: boolean | null;
  crowdinArticleDirectory?: (string | null) | CrowdinArticleDirectory;
  updatedAt: string;
  createdAt: string;
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "crowdin-article-directories".
 */
export interface CrowdinArticleDirectory {
  id: string;
  excludeLocales?: ('de_DE' | 'fr_FR')[] | null;
  name?: string | null;
  crowdinCollectionDirectory?: (string | null) | CrowdinCollectionDirectory;
  crowdinFiles?: (string | CrowdinFile)[] | null;
  parent?: (string | null) | CrowdinArticleDirectory;
  reference?: {
    createdAt?: string | null;
    updatedAt?: string | null;
    projectId?: number | null;
  };
  originalId?: number | null;
  directoryId?: number | null;
  updatedAt: string;
  createdAt: string;
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "crowdin-collection-directories".
 */
export interface CrowdinCollectionDirectory {
  id: string;
  name?: string | null;
  title?: string | null;
  collectionSlug?: string | null;
  reference?: {
    createdAt?: string | null;
    updatedAt?: string | null;
    projectId?: number | null;
  };
  originalId?: number | null;
  directoryId?: number | null;
  updatedAt: string;
  createdAt: string;
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "crowdin-files".
 */
export interface CrowdinFile {
  id: string;
  title?: string | null;
  field?: string | null;
  crowdinArticleDirectory?: (string | null) | CrowdinArticleDirectory;
  reference?: {
    createdAt?: string | null;
    updatedAt?: string | null;
    projectId?: number | null;
  };
  originalId?: number | null;
  directoryId?: number | null;
  revisionId?: number | null;
  name?: string | null;
  type?: ('json' | 'html') | null;
  path?: string | null;
  fileData?: {
    json?:
      | {
          [k: string]: unknown;
        }
      | unknown[]
      | string
      | number
      | boolean
      | null;
    html?: string | null;
  };
  updatedAt: string;
  createdAt: string;
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "localized-posts".
 */
export interface LocalizedPost {
  id: string;
  title?: string | null;
  author?: (string | null) | User;
  publishedDate?: string | null;
  category?: (string | null) | Category;
  tags?: (string | Tag)[] | null;
  content?:
    | {
        [k: string]: unknown;
      }[]
    | null;
  status?: ('draft' | 'published') | null;
  syncTranslations?: boolean | null;
  syncAllTranslations?: boolean | null;
  crowdinArticleDirectory?: (string | null) | CrowdinArticleDirectory;
  updatedAt: string;
  createdAt: string;
  _status?: ('draft' | 'published') | null;
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "users".
 */
export interface User {
  id: string;
  name?: string | null;
  updatedAt: string;
  createdAt: string;
  email: string;
  resetPasswordToken?: string | null;
  resetPasswordExpiration?: string | null;
  salt?: string | null;
  hash?: string | null;
  loginAttempts?: number | null;
  lockUntil?: string | null;
  password: string | null;
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "tags".
 */
export interface Tag {
  id: string;
  name?: string | null;
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "media".
 */
export interface Media {
  id: string;
  alt?: string | null;
  updatedAt: string;
  createdAt: string;
  url?: string | null;
  filename?: string | null;
  mimeType?: string | null;
  filesize?: number | null;
  width?: number | null;
  height?: number | null;
  focalX?: number | null;
  focalY?: number | null;
  sizes?: {
    thumbnail?: {
      url?: string | null;
      width?: number | null;
      height?: number | null;
      mimeType?: string | null;
      filesize?: number | null;
      filename?: string | null;
    };
    card?: {
      url?: string | null;
      width?: number | null;
      height?: number | null;
      mimeType?: string | null;
      filesize?: number | null;
      filename?: string | null;
    };
    tablet?: {
      url?: string | null;
      width?: number | null;
      height?: number | null;
      mimeType?: string | null;
      filesize?: number | null;
      filename?: string | null;
    };
  };
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "nested-field-collection".
 */
export interface NestedFieldCollection {
  id: string;
  title?: string | null;
  textField?: string | null;
  richTextField?:
    | {
        [k: string]: unknown;
      }[]
    | null;
  textareaField?: string | null;
  arrayField?:
    | {
        textField?: string | null;
        richTextField?:
          | {
              [k: string]: unknown;
            }[]
          | null;
        textareaField?: string | null;
        id?: string | null;
      }[]
    | null;
  layout?:
    | (
        | {
            textField?: string | null;
            richTextField?:
              | {
                  [k: string]: unknown;
                }[]
              | null;
            textareaField?: string | null;
            id?: string | null;
            blockName?: string | null;
            blockType: 'basicBlock';
          }
        | {
            textField?: string | null;
            richTextField?:
              | {
                  [k: string]: unknown;
                }[]
              | null;
            textareaField?: string | null;
            id?: string | null;
            blockName?: string | null;
            blockType: 'basicBlockNonLocalized';
          }
        | {
            richTextField?:
              | {
                  [k: string]: unknown;
                }[]
              | null;
            id?: string | null;
            blockName?: string | null;
            blockType: 'basicBlockRichText';
          }
        | {
            textField?: string | null;
            richTextField?:
              | {
                  [k: string]: unknown;
                }[]
              | null;
            textareaField?: string | null;
            id?: string | null;
            blockName?: string | null;
            blockType: 'basicBlockMixed';
          }
        | {
            title?: string | null;
            messages?:
              | {
                  title?: string | null;
                  message?:
                    | {
                        [k: string]: unknown;
                      }[]
                    | null;
                  id?: string | null;
                }[]
              | null;
            id?: string | null;
            blockName?: string | null;
            blockType: 'testBlockArrayOfRichText';
          }
      )[]
    | null;
  group?: {
    textField?: string | null;
    richTextField?:
      | {
          [k: string]: unknown;
        }[]
      | null;
    textareaField?: string | null;
  };
  tabOneTitle?: string | null;
  tabOneContent?:
    | {
        [k: string]: unknown;
      }[]
    | null;
  tabTwo: {
    tabTwoTitle?: string | null;
    tabTwoContent?:
      | {
          [k: string]: unknown;
        }[]
      | null;
  };
  syncTranslations?: boolean | null;
  syncAllTranslations?: boolean | null;
  crowdinArticleDirectory?: (string | null) | CrowdinArticleDirectory;
  updatedAt: string;
  createdAt: string;
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "policies".
 */
export interface Policy {
  id: string;
  title?: string | null;
  content?: {
    root: {
      type: string;
      children: {
        type: string;
        version: number;
        [k: string]: unknown;
      }[];
      direction: ('ltr' | 'rtl') | null;
      format: 'left' | 'start' | 'center' | 'right' | 'end' | 'justify' | '';
      indent: number;
      version: number;
    };
    [k: string]: unknown;
  } | null;
  group?: {
    array?:
      | {
          title?: string | null;
          content?: {
            root: {
              type: string;
              children: {
                type: string;
                version: number;
                [k: string]: unknown;
              }[];
              direction: ('ltr' | 'rtl') | null;
              format: 'left' | 'start' | 'center' | 'right' | 'end' | 'justify' | '';
              indent: number;
              version: number;
            };
            [k: string]: unknown;
          } | null;
          id?: string | null;
        }[]
      | null;
  };
  syncTranslations?: boolean | null;
  syncAllTranslations?: boolean | null;
  crowdinArticleDirectory?: (string | null) | CrowdinArticleDirectory;
  updatedAt: string;
  createdAt: string;
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "posts".
 */
export interface Post {
  id: string;
  title?: string | null;
  author?: (string | null) | User;
  publishedDate?: string | null;
  category?: (string | null) | Category;
  tags?: (string | Tag)[] | null;
  content?:
    | {
        [k: string]: unknown;
      }[]
    | null;
  status?: ('draft' | 'published') | null;
  updatedAt: string;
  createdAt: string;
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "localized-posts-with-condition".
 */
export interface LocalizedPostsWithCondition {
  id: string;
  title?: string | null;
  translateWithCrowdin?: boolean | null;
  author?: (string | null) | User;
  publishedDate?: string | null;
  category?: (string | null) | Category;
  tags?: (string | Tag)[] | null;
  content?:
    | {
        [k: string]: unknown;
      }[]
    | null;
  status?: ('draft' | 'published') | null;
  syncTranslations?: boolean | null;
  syncAllTranslations?: boolean | null;
  crowdinArticleDirectory?: (string | null) | CrowdinArticleDirectory;
  updatedAt: string;
  createdAt: string;
  _status?: ('draft' | 'published') | null;
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "payload-preferences".
 */
export interface PayloadPreference {
  id: string;
  user: {
    relationTo: 'users';
    value: string | User;
  };
  key?: string | null;
  value?:
    | {
        [k: string]: unknown;
      }
    | unknown[]
    | string
    | number
    | boolean
    | null;
  updatedAt: string;
  createdAt: string;
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "payload-migrations".
 */
export interface PayloadMigration {
  id: string;
  name?: string | null;
  batch?: number | null;
  updatedAt: string;
  createdAt: string;
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "localized-nav".
 */
export interface LocalizedNav {
  id: string;
  items: {
    label?: string | null;
    id?: string | null;
  }[];
  syncTranslations?: boolean | null;
  syncAllTranslations?: boolean | null;
  crowdinArticleDirectory?: (string | null) | CrowdinArticleDirectory;
  updatedAt?: string | null;
  createdAt?: string | null;
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "nav".
 */
export interface Nav {
  id: string;
  items: {
    label?: string | null;
    id?: string | null;
  }[];
  updatedAt?: string | null;
  createdAt?: string | null;
}
/**
 * This interface was referenced by `Config`'s JSON-Schema
 * via the `definition` "statistics".
 */
export interface Statistic {
  id: string;
  users?: {
    text?: string | null;
    number?: number | null;
  };
  countries?: {
    text?: string | null;
    number?: number | null;
  };
  syncTranslations?: boolean | null;
  syncAllTranslations?: boolean | null;
  crowdinArticleDirectory?: (string | null) | CrowdinArticleDirectory;
  _status?: ('draft' | 'published') | null;
  updatedAt?: string | null;
  createdAt?: string | null;
}


declare module 'payload' {
  export interface GeneratedTypes extends Config {}
}