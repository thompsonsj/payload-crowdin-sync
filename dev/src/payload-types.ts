/* tslint:disable */
/* eslint-disable */
/**
 * This file was automatically generated by Payload.
 * DO NOT MODIFY IT BY HAND. Instead, modify your source Payload config,
 * and re-run `payload generate:types` to regenerate this file.
 */

export interface Config {
  collections: {
    categories: Category;
    'localized-posts': LocalizedPost;
    'nested-field-collection': NestedFieldCollection;
    posts: Post;
    tags: Tag;
    users: User;
    'crowdin-files': CrowdinFile;
    'crowdin-collection-directories': CrowdinCollectionDirectory;
    'crowdin-article-directories': CrowdinArticleDirectory;
  };
  globals: {
    nav: Nav;
  };
}
export interface Category {
  id: string;
  name?: string;
}
export interface LocalizedPost {
  id: string;
  title?: string;
  author?: string | User;
  publishedDate?: string;
  category?: string | Category;
  tags?: string[] | Tag[];
  content?: {
    [k: string]: unknown;
  }[];
  status?: 'draft' | 'published';
  crowdinArticleDirectory?: string | CrowdinArticleDirectory;
  updatedAt: string;
  createdAt: string;
}
export interface User {
  id: string;
  name?: string;
  updatedAt: string;
  createdAt: string;
  email: string;
  resetPasswordToken?: string;
  resetPasswordExpiration?: string;
  salt?: string;
  hash?: string;
  loginAttempts?: number;
  lockUntil?: string;
  password?: string;
}
export interface Tag {
  id: string;
  name?: string;
}
export interface CrowdinArticleDirectory {
  id: string;
  name?: string;
  crowdinCollectionDirectory?: string | CrowdinCollectionDirectory;
  crowdinFiles?: string[] | CrowdinFile[];
  createdAt: string;
  updatedAt: string;
  originalId?: number;
  projectId?: number;
  directoryId?: number;
  excludeLocales?: ('de_DE' | 'fr_FR')[];
}
export interface CrowdinCollectionDirectory {
  id: string;
  name?: string;
  title?: string;
  collectionSlug?: string;
  createdAt: string;
  updatedAt: string;
  originalId?: number;
  projectId?: number;
  directoryId?: number;
}
export interface CrowdinFile {
  id: string;
  title?: string;
  field?: string;
  crowdinArticleDirectory?: string | CrowdinArticleDirectory;
  createdAt: string;
  updatedAt: string;
  originalId?: number;
  projectId?: number;
  directoryId?: number;
  revisionId?: number;
  name?: string;
  type?: string;
  path?: string;
  fileData?: {
    json?:
      | {
          [k: string]: unknown;
        }
      | unknown[]
      | string
      | number
      | boolean
      | null;
    html?: string;
  };
}
export interface NestedFieldCollection {
  id: string;
  textField?: string;
  richTextField?: {
    [k: string]: unknown;
  }[];
  textareaField?: string;
  arrayField?: {
    textField?: string;
    richTextField?: {
      [k: string]: unknown;
    }[];
    textareaField?: string;
    id?: string;
  }[];
  layout?: (
    | {
        textField?: string;
        richTextField?: {
          [k: string]: unknown;
        }[];
        textareaField?: string;
        id?: string;
        blockName?: string;
        blockType: 'basicBlock';
      }
    | {
        richTextField?: {
          [k: string]: unknown;
        }[];
        id?: string;
        blockName?: string;
        blockType: 'basicBlockRichText';
      }
    | {
        textField?: string;
        richTextField?: {
          [k: string]: unknown;
        }[];
        textareaField?: string;
        id?: string;
        blockName?: string;
        blockType: 'basicBlockMixed';
      }
    | {
        title?: string;
        messages?: {
          title?: string;
          message?: {
            [k: string]: unknown;
          }[];
          id?: string;
        }[];
        id?: string;
        blockName?: string;
        blockType: 'testBlockArrayOfRichText';
      }
  )[];
  group?: {
    textField?: string;
    richTextField?: {
      [k: string]: unknown;
    }[];
    textareaField?: string;
  };
  tabOneTitle?: string;
  tabOneContent?: {
    [k: string]: unknown;
  }[];
  tabTwo: {
    tabTwoTitle?: string;
    tabTwoContent?: {
      [k: string]: unknown;
    }[];
  };
  crowdinArticleDirectory?: string | CrowdinArticleDirectory;
  updatedAt: string;
  createdAt: string;
}
export interface Post {
  id: string;
  title?: string;
  author?: string | User;
  publishedDate?: string;
  category?: string | Category;
  tags?: string[] | Tag[];
  content?: {
    [k: string]: unknown;
  }[];
  status?: 'draft' | 'published';
  updatedAt: string;
  createdAt: string;
}
export interface Nav {
  id: string;
  items: {
    label?: string;
    id?: string;
  }[];
  crowdinArticleDirectory?: string | CrowdinArticleDirectory;
  updatedAt?: string;
  createdAt?: string;
}
