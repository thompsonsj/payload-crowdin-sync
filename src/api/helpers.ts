import { Payload } from "payload";
import { IcrowdinFile } from "./payload-crowdin-sync/files";

/**
 * get Crowdin Article Directory for a given documentId
 *
 * The Crowdin Article Directory is associated with a document,
 * so is easy to retrieve. Use this function when you only have
 * a document id.
 */
export async function getArticleDirectory(
  documentId: string,
  payload: Payload,
  allowEmpty?: boolean
) {
  // Get directory
  const crowdinPayloadArticleDirectory = await payload.find({
    collection: "crowdin-article-directories",
    where: {
      name: {
        equals: documentId,
      },
    },
  });
  if (crowdinPayloadArticleDirectory.totalDocs === 0 && allowEmpty) {
    // a thrown error won't be reported in an api call, so console.log it as well.
    console.log(`No article directory found for document ${documentId}`);
    throw new Error(
      "This article does not have a corresponding entry in the  crowdin-article-directories collection."
    );
  }
  return crowdinPayloadArticleDirectory
    ? crowdinPayloadArticleDirectory.docs[0]
    : undefined;
}

export async function getFile(
  name: string,
  crowdinArticleDirectoryId: string,
  payload: Payload
): Promise<any> {
  const result = await payload.find({
    collection: "crowdin-files",
    where: {
      field: { equals: name },
      crowdinArticleDirectory: {
        equals: crowdinArticleDirectoryId,
      },
    },
  });
  return result.docs[0];
}

export async function getFiles(
  crowdinArticleDirectoryId: string,
  payload: Payload
): Promise<any> {
  const result = await payload.find({
    collection: "crowdin-files",
    limit: 10000,
    where: {
      crowdinArticleDirectory: {
        equals: crowdinArticleDirectoryId,
      },
    },
  });
  return result.docs;
}

export async function getFileByDocumentID(
  name: string,
  documentId: string,
  payload: Payload
): Promise<IcrowdinFile> {
  const articleDirectory = await getArticleDirectory(documentId, payload);
  return getFile(name, `${articleDirectory?.id}`, payload);
}

export async function getFilesByDocumentID(
  documentId: string,
  payload: Payload
): Promise<IcrowdinFile[]> {
  const articleDirectory = await getArticleDirectory(documentId, payload);
  if (!articleDirectory) {
    // tests call this function to make sure files are deleted
    return [];
  }
  const files = await getFiles(`${articleDirectory.id}`, payload);
  return files;
}
