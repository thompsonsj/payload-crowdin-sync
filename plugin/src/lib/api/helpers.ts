import payload, { Payload } from "payload";
import { CrowdinArticleDirectory, CrowdinCollectionDirectory, CrowdinFile } from "../payload-types";
import { payloadCrowdinSyncTranslationsApi } from "./payload-crowdin-sync/translations";
import { PluginOptions, isCollectionOrGlobalConfigObject, isCollectionOrGlobalConfigSlug } from "../types";
import { ReadableStreamDefaultController } from "stream/web";
import { Plugin } from "payload/config";

/**
 * get Crowdin Article Directory for a given documentId
 *
 * The Crowdin Article Directory is associated with a document,
 * so is easy to retrieve. Use this function when you only have
 * a document id.
 */
export async function getArticleDirectory(
  documentId: string,
  payload: Payload,
  allowEmpty?: boolean
) {
  // Get directory
  const crowdinPayloadArticleDirectory = await payload.find({
    collection: "crowdin-article-directories",
    where: {
      name: {
        equals: documentId,
      },
    },
  });
  if (crowdinPayloadArticleDirectory.totalDocs === 0 && allowEmpty) {
    // a thrown error won't be reported in an api call, so console.log it as well.
    console.log(`No article directory found for document ${documentId}`);
    throw new Error(
      "This article does not have a corresponding entry in the  crowdin-article-directories collection."
    );
  }
  return crowdinPayloadArticleDirectory
    ? crowdinPayloadArticleDirectory.docs[0]
    : undefined;
}

export async function getFile(
  name: string,
  crowdinArticleDirectoryId: string,
  payload: Payload
): Promise<any> {
  const result = await payload.find({
    collection: "crowdin-files",
    where: {
      field: { equals: name },
      crowdinArticleDirectory: {
        equals: crowdinArticleDirectoryId,
      },
    },
  });
  return result.docs[0];
}

export async function getFiles(
  crowdinArticleDirectoryId: string,
  payload: Payload
): Promise<any> {
  const result = await payload.find({
    collection: "crowdin-files",
    limit: 10000,
    where: {
      crowdinArticleDirectory: {
        equals: crowdinArticleDirectoryId,
      },
    },
  });
  return result.docs;
}

export async function getFileByDocumentID(
  name: string,
  documentId: string,
  payload: Payload
): Promise<CrowdinFile> {
  const articleDirectory = await getArticleDirectory(documentId, payload);
  return getFile(name, `${articleDirectory?.id}`, payload);
}

export async function getFilesByDocumentID(
  documentId: string,
  payload: Payload
): Promise<CrowdinFile[]> {
  const articleDirectory = await getArticleDirectory(documentId, payload);
  if (!articleDirectory) {
    // tests call this function to make sure files are deleted
    return [];
  }
  const files = await getFiles(`${articleDirectory.id}`, payload);
  return files;
}

interface IupdatePayloadTranslation {
  articleDirectoryId: string
  pluginOptions: PluginOptions,
  payload: Payload
  /** store translations into a draft */
  draft?: boolean
  /** prevent database changes */
  dryRun?: boolean
  /** override article directory exclude locales */
  excludeLocales?: string[]
}

export async function updatePayloadTranslation({
  articleDirectoryId,
  pluginOptions,
  payload,
  draft,
  dryRun,
  excludeLocales,
}: IupdatePayloadTranslation) {
  // get article directory
  const articleDirectory = await payload.findByID({
    id: articleDirectoryId,
    collection: "crowdin-article-directories",
  });
  // is this a global or a collection?
  const global =
    (articleDirectory['crowdinCollectionDirectory'] as CrowdinCollectionDirectory)?.collectionSlug as string === "globals";
  // get an instance of our translations api
  const translationsApi = new payloadCrowdinSyncTranslationsApi(
    pluginOptions,
    payload
  );
  try {
    const translations = await translationsApi.updateTranslation({
      documentId: !global ? articleDirectory["name"] as string : ``,
      collection: global
        ? articleDirectory['name'] as string
        : (articleDirectory["crowdinCollectionDirectory"] as CrowdinCollectionDirectory)?.collectionSlug as string,
      global,
      draft,
      dryRun,
      excludeLocales: excludeLocales || articleDirectory["excludeLocales"] as string[] || [],
    });
    return {
      status: 200,
      ...translations
    };
  } catch (error) {
    return {
      status: 400,
      error,
    }
  }
}

export const isCrowdinActive = ({
  doc,
  slug,
  global,
  pluginOptions
}: {
  doc: any,
  slug: string
  global: boolean
  pluginOptions: PluginOptions
}) => {
  if (!slug) {
    return false
  }

  if (global) {
    if (!pluginOptions.globals) {
      return true
    }

    const matchingGlobalConfig = pluginOptions.globals.find(config => {
      if (isCollectionOrGlobalConfigObject(config) && config.slug === slug) {
        if (typeof config.condition !== 'undefined') {
          return config.condition({ doc })
        }
        return true
      }
      if (isCollectionOrGlobalConfigSlug(config)) {
        return config === slug
      }
      return false
    })

    if (typeof matchingGlobalConfig !== 'undefined') {
      return true
    }
  }

  if (!pluginOptions.collections) {
    return true
  }

  const matchingCollectionConfig = pluginOptions.collections.find(config => {
    if (isCollectionOrGlobalConfigObject(config) && config.slug === slug) {
      if (typeof config.condition !== 'undefined') {
        return config.condition({ doc })
      }
      return true
    }
    if (isCollectionOrGlobalConfigSlug(config)) {
      return config === slug
    }
    return false
  })

  if (typeof matchingCollectionConfig !== 'undefined') {
    return true
  }

  return false
}
