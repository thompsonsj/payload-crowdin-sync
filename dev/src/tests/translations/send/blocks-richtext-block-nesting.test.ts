import { getFiles, getLexicalFieldArticleDirectories } from "payload-crowdin-sync";
import nock from "nock";
import { mockCrowdinClient } from "payload-crowdin-sync";
import { pluginConfig } from "../../helpers/plugin-config"

import { fixture } from "./__fixtures__/block-richtext-block-nesting.fixture";

import { initPayloadInt } from '../../helpers/initPayloadInt'
import type { Payload } from 'payload'
import { CrowdinArticleDirectory } from "@/payload-types";

let payload: Payload

// const fixtureHtmlFr = '<div class="rich-text"><p>Text in a Lexical field inside a block in the <strong>block</strong> of a <em>Lexical field</em> of a <strong>Lexical field</strong> of a <strong>Lexical field</strong>.</p></div>'

/**
 * Test translations
 *
 * Ensure translations are retrieved, compared, and
 * stored as expected.
 */

const pluginOptions = pluginConfig()
const mockClient = mockCrowdinClient(pluginOptions)

/**
 * payload.config.custom-serializers.ts required for htmlToSlate.
 * 
 * Plugin options passed directly to payloadCrowdinSyncTranslationsApi - this API is initialized through `server.ts` at the moment. Can change this in the future to use internal functions.
 */
describe("Translations", () => {
  beforeAll(async () => {
    const initialized = await initPayloadInt()
        ;({ payload } = initialized as {
          payload: Payload
        })
  });

  afterEach((done) => {
    if (!nock.isDone()) {
      throw new Error(
        `Not all nock interceptors were used: ${JSON.stringify(
          nock.pendingMocks()
        )}`
      );
    }
    nock.cleanAll()
    done()
  })

  afterAll(async () => {
    if (typeof payload.db.destroy === 'function') {
      await payload.db.destroy()
    }
  });

  describe("fn: updateTranslation - custom serializer", () => {
    it("updates a Payload article with a `richText` field translation retrieved from Crowdin restoring table structure", async () => {
      const fileId = 4674

      nock("https://api.crowdin.com")
        .post(
          `/api/v2/projects/${pluginOptions.projectId}/directories`
        )
        .times(3)
        .reply(200, mockClient.createDirectory({}))
        .post(
          `/api/v2/storages`
        )
        .reply(200, mockClient.addStorage())
        .post(
          `/api/v2/projects/${pluginOptions.projectId}/files`
        )
        .reply(200, mockClient.createFile({
          fileId,
        }))
        /* .post(
          `/api/v2/projects/${pluginOptions.projectId}/translations/builds/files/${fileId}`,
          {
            targetLanguageId: 'fr',
          }
        )
        .reply(200, mockClient.buildProjectFileTranslation({
          url: `https://api.crowdin.com/api/v2/projects/${pluginOptions.projectId}/translations/builds/${fileId}/download?targetLanguageId=fr`
        }))
        .get(
          `/api/v2/projects/${pluginOptions.projectId}/translations/builds/${fileId}/download`
        )
        .query({
          targetLanguageId: 'fr',
        })
        .reply(200, fixtureHtmlFr,
        ) */;
      const create = await payload.create({
        collection: "nested-field-collection",
        data: {
          title: "Non-localized title",
          layout: fixture,
        },
      });
      // Crowdin file
      const post = await payload.findByID({
        collection: "nested-field-collection",
        id: create.id,
      })
      const files = await getFiles( (post.crowdinArticleDirectory as CrowdinArticleDirectory)?.id, payload)
      expect(files.length).toBe(1);
      const file = files[0];
      expect(file.fileData?.html).toMatchSnapshot();
      expect(file.fileData?.sourceBlocks).toMatchSnapshot();
      const childDirectories = await getLexicalFieldArticleDirectories({
        payload,
        parent: post.crowdinArticleDirectory,
      })
      expect(childDirectories.length).toBe(1);
      console.log(childDirectories)
      const childDirectory = childDirectories[0]
      expect(childDirectory.name).toEqual(`lex.layout.688dd685258ffa0224ea6daa.basicBlockLexical.content`)
      const childDirectoryFiles = await getFiles(childDirectory.id, payload);
      // here's the issue - Crowdin files are not created for the child directory. Is this a simple localization fix?
      expect(childDirectoryFiles.length).toBeGreaterThan(0)
      /* const regexp = /\w{8}-\w{4}-\w{4}-\w{4}-\w{12}/g;
      regexp.lastIndex = 1;
      const str = file.docs[0].fileData?.html || ``;
      const uuids = Array.from(str.matchAll(regexp), (m) => m[0]);
      let fixtureHtmlWithUuids = fixtureHtml
      uuids.forEach((uuid, index) => {
        fixtureHtmlWithUuids = fixtureHtmlWithUuids.replace(`<uuid[${index}]>`, uuid)
      })
      expect(file.docs[0].fileData?.html).toEqual(fixtureHtmlWithUuids);
      const translationsApi = new payloadCrowdinSyncTranslationsApi(
        pluginOptions,
        payload,
      );
      await translationsApi.updateTranslation({
        documentId: `${post.id}`,
        collection: "policies",
        dryRun: false,
      });
      // retrieve translated post from Payload
      const frResult = await payload.findByID({
        collection: "policies",
        id: post.id,
        locale: "fr_FR",
      });
      
      expect(frResult["content"]).toEqual(fixtureFr); */
    });
  });
});
