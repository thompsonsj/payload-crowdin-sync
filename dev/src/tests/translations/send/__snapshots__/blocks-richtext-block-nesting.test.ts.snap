// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Translations fn: updateTranslation - custom serializer updates a Payload article with a \`richText\` field translation retrieved from Crowdin restoring table structure 1`] = `
{
  "blocks": {
    "688dd71aa950db7862106963": {
      "featureList": {
        "items": {
          "688dd71d258ffa0224ea6dac": {
            "text": "Text in an array field in the 'Feature List' block. This block is embedded in a parent Lexical field within the 'Highlight' block.",
          },
          "688dd71f258ffa0224ea6dae": {
            "text": "The Payload Crowdin Sync plugin can parse complex field structures including blocks nested in Lexical fields nested in blocks nested in ...etc.",
          },
        },
      },
    },
  },
}
`;
